Architectural overview

antalszava opened the pull request:
**Description of the Change:**
Extends the documentation page with an architectural overview page that:

* Provides a high-level overview of the components of PennyLane, the
  interaction among them and related links
* Contains information and related links on key design details

**Benefits:**
* Provides a single page where further parts of the documentation related to each component are easily accessible
* Provides a place where design details can be added

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+Queuing of operators
+********************
+
+In PennyLane, the construction of quantum gates is separated from the specific
+quantum node (:class:`~.BaseQNode`) that they belong to. However, including
+logic for this when creating an instance of :class:`~.Operator` does not align
+with the current architecture. Therefore, there is a need to use a high-level
+object that holds information about the relationship between quantum gates and
+a quantum node.
+
+The :class:`~.QueuingContext` class realizes this by providing access to the current
+QNode.  Furthermore, it provides the flexibility to have multiple objects
+record the creation of quantum gates.
+
+The ``QueuingContext`` class both acts as the abstract base class for all
+classes that expose a queue for Operations (so-called contexts), as well as the
+interface to said queues. The active contexts contain maximally one QNode and
+an arbitrary number of other contexts like the :class:`~.OperationRecorder`.
```
mostly preferred placing pointer links such that we only have to maintain a text at one place. This one is an exception as the text is also part of the docstring of the QueuingContext class. (was unsure how this could be referenced easily there :thinking:)

antalszava commented:
Hi @josh146, let me know if you'd suggest to change any part of the structure here and/or add new sections. A thing I was unsure of is the list of devices: these could be organized into another page if need be.

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+
+Each ``QNode`` represents the quantum circuit by building a
+:class:`~.CircuitGraph` instance, but the way differentiation is done is custom
+to the differentiation method offered by the ``QNode``.
```
In addition, there is a distinction between differentiable QNodes (those that inherit from `JacobianQNode`) and 'non-differentiable' QNodes (`BaseQNode`, `PasshthruQNode`).

co9olguy replied:
Maybe worthwhile to start off with an overview/reminder of the core functionality provided by PennyLane (e.g., circuit evaluation & gradients on diverse quantum hardware, all connected to classical machine learning libraries), and use that as a motivator for what role all of the components play?

antalszava replied:
Thanks, great ideas! Extended the introduction with ideas on the core functionality of PL and added a list on diff rules

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+interfaces.
+
+We refer to the :ref:`intro_interfaces` page for a more in-depth introduction
+and a list of available interfaces.
```
I think some additional developer details regarding the interfaces should go here (e.g., that an interface is simply a 'wrapper' around _differentiable/Jacobian_ QNodes such that the the QNode's quantum gradient rules are registered to the ML interface via a custom gradient class/function.

antalszava replied:
That a great point, included it!

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+We refer to the :ref:`intro_interfaces` page for a more in-depth introduction
+and a list of available interfaces.
+
+Optimizers
+**********
+
+Optimizers are objects which can be used to automatically update the parameters
+of a quantum or hybrid machine learning model.
+
+We refer to the :ref:`intro_ref_opt` page for a more in-depth introduction
+and a list of available optimizers.
```
As above, I'm thinking we remove this, as Optimizers are really 'core'.

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+
+The following are key design details related to how PennyLane works internally.
+
+Queuing of operators
+********************
```
I think this is great to include! However, we could potentially include more specifics; e.g.,

* Operations are queued on initialization, via a call to `Operation.queue()`

* Already instantiated operations cannot be re-queued unless `op.queue()` is manually called within the queueing context.

* `Observables` do not queue themselves; the _measurement functions_ `qml.expval`, `qml.var` etc. do the queueing. Hence, observables _can_ be instantiated before queueing.

It also seems a good place to mention the `OperationRecorder`.

antalszava replied:
Great points, added them! :100:

co9olguy commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+device.
+
+Each ``QNode`` represents the quantum circuit by building a
+:class:`~.CircuitGraph` instance, but the way differentiation is done is custom
```
This might be confusing to readers, since (I don't believe) CircuitGraph is not mentione elsewhere in the doc

antalszava replied:
included further details

co9olguy commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+Queuing of operators
+********************
+
+In PennyLane, the construction of quantum gates is separated from the specific
+quantum node (:class:`~.BaseQNode`) that they belong to. However, including
+logic for this when creating an instance of :class:`~.Operator` does not align
+with the current architecture. Therefore, there is a need to use a high-level
```
I got a bit confused reading this paragraph. I imagine a reader might not understand everything that is mentioned

antalszava replied:
rephrased it, and hopefully, the end of the new `Operators` section would also serve as a good basis for this section

co9olguy commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+Components of PennyLane
+#######################
+
+.. image:: architecture_diagram.png
+    :width: 800px
```
yeah, this one is quite old, and doesn't do that great a job. For instance, the "Quantum device" should really be a picture of a device, not a picture of a circuit

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+
+For further details on qnodes, the mutability of qnodes we refer to
+:class:`~.BaseQNode` and for a list of qnodes to :ref:`qml_qnodes`.
+
```
Added!

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+Qubit based devices can use shared utilities by using the
+:class:`~.QubitDevice`.
+
+:html:`<div class="state_vec_list" id="aside1"><a data-toggle="collapse" data-parent="#aside1" href="#content1" class="collapsed"><p class="first admonition-title">Statevector simulators (click to expand) <i class="fas fa-chevron-circle-down"></i></p></a><div id="content1" class="collapse" data-parent="#aside1" style="height: 0px;">`
```
True, I was also hesitating about this! Removed it from here

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,228 @@
+``Device`` class as their parent class.
+
+A ``QNode`` is another important component in PennyLane (later detailed), they
+serve as the abstraction for quantum circuits to be run on a device.
```
True, removed it!

antalszava commented:
Thank you for the comments @co9olguy @josh146! I'd hope that I addressed all of your comments, the changes include:
* adding a more general introduction for the page on core functionalities of PL
*  adding a list of QNode diff types
* adding an operators section which also prepares the section on `queuing`

Further to your suggestions, a new diagram is to be added.

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+
+The integration between classical and quantum computations is encompassed by
+interfaces. Differentiable and Jacobian ``QNodes`` have interfacing functions
+that provide a 'wrapper' around QNodes such. These wrappers further transform
+the ``QNode`` such that the quantum gradient rules of the QNodes are registered
+to the machine learning interface via a custom gradient class or function.
```
I suppose we could go into further detail here with what exactly an interface function does.

* It wraps the QNode, returning a QNode that accepts and returns the core data structure of the classical machine learning library (e.g., a TF tensor, Torch tensor, Autograd NumPy array, etc).

* It _unwraps_ the input data structures to simple NumPy arrays, so that the quantum device can execute the
  users quantum function.

* It registers the `QNode.jacobian()` method as a custom gradient method, so that the machine learning library can 'backpropagate' across the QNode, when integrated into a classical computation.

@co9olguy has a nice diagram that represents the unwrapping and evaluation performed by interfaces :)

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+Together with ``Operator``, these classes serve as base classes for quantum
+operators.
+
+Certain operators can serve as both quantum gates and observables (e.g.
+:class:`~.PauliZ`, :class:`~.PauliX`, etc.). Such classes inherit from both
+``Operation`` and ``Observable`` classes.
```
Nice!

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+:class:`~.PauliZ`, :class:`~.PauliX`, etc.). Such classes inherit from both
+``Operation`` and ``Observable`` classes.
+
+Quantum operator instances, in general, are used to build quantum functions
+which are used within ``QNode`` instances. Users can define such quantum
```
```suggestion
Quantum operators are used to build quantum functions
which are evaluated by a ``QNode`` on a bound device. Users can define such quantum
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+
+Quantum operator instances, in general, are used to build quantum functions
+which are used within ``QNode`` instances. Users can define such quantum
+functions by creating regular Python functions and incorporating ``Operator``
+instances.
```
```suggestion
functions by creating regular Python functions and instantiating ``Operator``
instances in temporal order, one per line.
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+        qml.PauliX(0)
+        return qml.expval(qml.PauliZ(0))
+
+This syntax of PennyLane results in the Operator instances not being "recorded"
+by the Python function. What is more, there is actually *no composition logic*
+in between the ``QNode`` that is being created and the ``Operator`` instances
+(``qml.PauliX(0)`` and ``qml.PauliZ(0)``), as there are merely class
+instantiations happening upon calling the ``circuit`` quantum function.
```
This paragraph is a bit unclear to me 🤔

antalszava replied:
I've reworded it a bit, do you think it's better or would there be further things that are unclear?

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+
+The :class:`~.QueuingContext` class realizes this by providing access to the
+current QNode using the concept of a Python context manager. This is achieved
+by inheriting the ``BaseQNode`` class from ``QueuingContext``.  Furthermore, it
```
```suggestion
by subclassing the ``BaseQNode`` class from ``QueuingContext``.  Furthermore, it
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+Details of queueing and specific operators can be described as follows:
+
+* ``Operators`` in a quantum function are queued on initialization.  This
+  happens via a call to :meth:`Operator.queue` which then interacts with the
```
```suggestion
  happens via a call to :meth:`.Operator.queue` which then interacts with the parent
```

antalszava replied:
The `QueuingContext` is not a parent to the `Operator` class, but a parent to `BaseQNode`

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+*********
+
+Circuit parameters in PennyLane are tracked and updated using
+:class:`~.Variable`. They play a key role in the evaluation of a ``QNode`` as
+the symbolic parameters are substituted with numeric values.
```
```suggestion
:class:`~.Variable`. They play a key role in the evaluation of ``QNode`` gradients, as
the symbolic parameters are substituted with numeric values. The ``Variable`` class plays
an important role in book-keeping, allowing PennyLane to keep track of which parameters are
used in which operations, and automatically perform the product and chain rule where required.
```

co9olguy commented:
Just as a high-level comment, I'm a bit worried that this is one of those things that might easily become out of date (and people will forget to update it when things change), like the plugin-template repository.

To me, the source code is the source of truth. While we can provide high-level pictures of some basic functionality, pre v1.0, even that is not guaranteed to be stable. For these reasons, I would hesitate to go into too much fine-grained details in such a architecture page

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,238 @@
+
+An interface integrates QNodes with external libraries by the following:
+
+* It wraps the QNode, returning a QNode that accepts and returns the core data
+structure of the classical machine learning library (e.g., a TF tensor, Torch
+tensor, Autograd NumPy array, etc).
+* It unwraps the input data structures to simple NumPy arrays, so that the
+quantum device can execute the user's quantum function.
+* It registers the ``QNode.jacobian()`` method as a custom gradient method, so that
+the machine learning library can 'backpropagate' across the QNode, when
+integrated into a classical computation.
```
```suggestion
* It wraps the QNode, returning a QNode that accepts and returns the core data
  structure of the classical machine learning library (e.g., a TF tensor, Torch
  tensor, Autograd NumPy array, etc).

* It unwraps the input data structures to simple NumPy arrays, so that the
  quantum device can execute the user's quantum function.

* It registers the ``QNode.jacobian()`` method as a custom gradient method, so that
  the machine learning library can 'backpropagate' across the QNode, when
  integrated into a classical computation.
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,238 @@
+        qml.PauliX(0)
+        return qml.expval(qml.PauliZ(0))
+
+This syntax of PennyLane results in the Operator instances not being "recorded"
+by the ``QNode``. This is so because there is *no association or composition
+logic* in between the ``QNode`` that is being created and the ``Operator``
+instances: the ``QNode`` does not *contain* any instances of quantum operators.
+By executing ``circuit``, instances of the ``qml.PauliX`` and ``qml.PauliZ``
+classes are created, but they are not associated with the ``QNode`` itself.
+This creates the problem of quantum functions (``circuit``) not being able to
+keep track of which quantum operations and observables are being applied.
```
I still find this paragraph a bit difficult to read 🤔 Perhaps it makes sense to remove it altogether, and simply go straight into describing queuing and circuit construction?

Per @co9olguy's comment, perhaps it makes sense to significantly reduce the details here (including the next subsection), and just outline what is happening in bullet point form. E.g.,

* QNode circuit construction doesn't happen until the QNode is evaluated.

* On QNode evaluation, the quantum function is executed.

  - Operators are queued to the QNode on instantiation, by having `Operator.__init__` call the `Operator.queue()` method. The operators themselves queue themselves to the surrounding :class:`~.QueuingContext`.

  - Measurement functions such as `qml.expval()` are responsible for queuing observables.

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,238 @@
+an important role in book-keeping, allowing PennyLane to keep track of which parameters are
+used in which operations, and automatically perform the product and chain rule where required.
+
+We refer to the :ref:`qml_variable` page for a more in-depth description of how
```
```suggestion
We refer to the :doc:`/code/qml_variable` page for a more in-depth description of how
```

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,204 @@
+* :class:`~.JacobianQNode`: finite differences
+* :class:`~.DeviceJacobianQNode`: device based gradient rule
+* :class:`~.PassthruQNode`: classical backpropagation
+* :class:`~.ReversibleQNode`: reversible backpropagation
```
@josh146 keeping this page minimalistic, I'm having second thoughts on having this list in here. With new QNodes introduced, this will also have to be extended. How about we keep the reference to the list of QNodes?

antalszava commented:
Thanks for the comments!

* Included the new diagram
* Minimized further the text especially on queuing

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+:math:`f(x;\theta)=R^m\rightarrow R^n` that is executed using quantum
+information processing on a quantum device.
+
+Apart from incorporating quantum functions, QNodes also offer custom
+quantum differentiation rules. Using the so-called `parameter-shift rules
+<https://pennylane.ai/qml/glossary/parameter_shift.html>`__, many quantum
+functions can be expressed through the linear combination of other quantum
+functions. As these rules allow quantum gradients to be obtained from
+QNodes, hybrid computations may include QNodes as part of training deep
+learnings models.
```
A couple of suggestions here:

> Apart from **encapsulating** quantum functions, QNodes also **provide** ~custom~
> quantum differentiation rules. **Examples include** the parameter-shift rule, **where the derivative
> of** quantum functions can be expressed **by a** linear combination of **the** ~other~ quantum
> function. As these rules allow quantum gradients to be obtained from
> QNodes, hybrid computations may include QNodes as part of training deep
> learnings models.

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+QNodes, hybrid computations may include QNodes as part of training deep
+learnings models.
+
+These QNode types are available to users through the :func:`~.qnode` decorator by
+passing the user-facing ``diff_method`` option. This decorator then uses the
+:func:`~.QNode` constructor function to create the specific type of qnode based on
+the device, interface, and quantum function. If ``diff_method`` option is not
+provided, the QNode constructor function attempts to determine the ``"best"``
+differentiation method, based on the available device and interface.
```
Minor suggestion: might be good to first introduce the QNode function before mentioning `diff_method`?

```suggestion
Users don't typically instantiate QNodes directly---instead, the :func:`~.qnode` decorator or
:func:`~.QNode` constructor function automates the process of creating a QNode from a provided
quantum function and device. The constructor attempts to determine the ``"best"`` QNode
subclass/differentiation method for the provided device and interface. For more fine-grained control,
the differentiation method can be specified directly via the ``diff_method`` option.
```
(feel free to modify!)

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+provided, the QNode constructor function attempts to determine the ``"best"``
+differentiation method, based on the available device and interface.
+
+A widespread representation of quantum circuits is by creating a `Directed
```
```suggestion
A common representation of quantum circuits is by creating a `Directed
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+
+The integration between classical and quantum computations is encompassed by
+interfaces. QNodes that provide black-box gradient rules are 'wrapped' by an interface function.
+that provide a 'wrapper' around QNodes such. These wrappers further transform
```
```suggestion
These wrappers further transform
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+the ``QNode`` such that the quantum gradient rules of the QNodes are registered
+to the machine learning interface via a custom gradient class or function.
+
+An interface integrates QNodes with external libraries by the following:
```
```suggestion
Typically, an interface integrates QNodes with external libraries as follows:
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+Quantum operators are incorporated by the :class:`~.Operator` class which
+contains basic information about the operator (e.g. number of parameters,
+number of wires it acts on, etc.) and further convenience methods (e.g.
+:attr:`~.Operator.matrix`, :attr:`~.Operator.eigvals`.
```
```suggestion
:attr:`~.Operator.matrix`, :attr:`~.Operator.eigvals`).
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+call the :meth:`Operator.queue` method. The operators themselves queue themselves to
+the surrounding :class:`~.QueuingContext`.
+
+Measurement functions such as :meth:`qml.expval` are responsible for queuing observables.
```
```suggestion
Measurement functions such as :func:`~.expval` are responsible for queuing observables.
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+QNode that they belong to. QNode circuit construction happens only when the
+QNode is evaluated. On QNode evaluation, the quantum function is executed.
+
+Operators are queued to the QNode on instantiation, by having :meth:`Operator.__init__`
+call the :meth:`Operator.queue` method. The operators themselves queue themselves to
```
```suggestion
Operators are queued to the QNode on instantiation, by having :meth:`.Operator.__init__`
call the :meth:`.Operator.queue` method. The operators themselves queue themselves to
```

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+A plugin system makes the framework compatible with any gate-based quantum
+simulator or hardware.
+
+Using PennyLane, quantum computing devices (``Devices``) can be used to
+evaluate quantum nodes (``QNodes``) and to return statistics of the results. To
+process the classical information obtained, ``Interfaces`` allow using
+accelerated machine learning libraries.
```
agreed!

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+information processing on a quantum device.
+
+Apart from incorporating quantum functions, QNodes also offer custom
+quantum differentiation rules. Using the so-called `parameter-shift rules
+<https://pennylane.ai/qml/glossary/parameter_shift.html>`__, many quantum
```
really cool, thank you!

antalszava commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,196 @@
+Components of PennyLane
+#######################
+
+.. image:: architecture_diagram.png
```
thanks for catching that!

thisac commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,195 @@
+namespace using Setuptools. Once registered, the device can be instantiated using the :func:`~.device`
+loader function.
+
+A Python package that registers one or more PennyLane device is known as a *plugin*. For more details
```
```suggestion
A Python package that registers one or more PennyLane devices is known as a *plugin*. For more details
```

thisac commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,195 @@
+Architectural overview
+======================
+
+PennyLane allows optimization and machine learning of quantum and hybrid
```
Is it correct to say that `PennyLane allows machine learning of computations`? I would probably reformulate this sentence. :thinking:

antalszava replied:
That's a fair point! I've reformulated it

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,195 @@
+Architectural overview
+======================
+
+PennyLane is a framework for optimization and machine learning of quantum and
+hybrid quantum-classical computations. The library provides a unified
+architecture for near-term quantum computing devices, supporting various
+quantum information paradigms.
```
Really nice intro!

antalszava replied:
as a matter of fact took some help from the [abstract](https://arxiv.org/pdf/1811.04968.pdf) here :smile:

josh146 replied:
😆

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,195 @@
+quantum circuits in a way that is compatible with classical techniques such as
+backpropagation. PennyLane thus extends the automatic differentiation
+algorithms common in optimization and machine learning to include quantum and
+`hybrid computations
+<https://pennylane.ai/qml/glossary/hybrid_computation.html#backpropagation-through-hybrid-computations>`__.
```
```suggestion
:doc:`hybrid computations <glossary/hybrid_computation>`.
```

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,195 @@
+:class:`~.CircuitGraph` instance.
+
+For further details on QNodes, and a full list of QNodes with their custom
+differentiation rule, refer to the :doc:`/code/qml_qnodes` module.
```
:+1:

josh146 commented on doc/development/guide/architecture.rst:
```diff
@@ -0,0 +1,195 @@
+Typically, an interface integrates QNodes with external libraries as follows:
+
+* It wraps the QNode, returning a QNode that accepts and returns the core data
+  structure of the classical machine learning library (e.g., a TF tensor, Torch
```
```suggestion
  structure of the classical machine learning library (e.g., a TensorFlow or PyTorch
```
